File: Solutions/804_Unique_Morse_Code_Words/804_Unique_Morse_Code_Words.cs
public class Solution {
    public int UniqueMorseRepresentations(string[] words) {
        var map = new string[]{".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};
        var res = new HashSet<string>();
        foreach(var word in words){
            var sb = new StringBuilder();
            foreach(var c in word.ToCharArray()){
                sb.Append(map[c-'a']);
            }
            res.Add(sb.ToString());
        }
        
        return res.Count;
    }
}

//Third fastest
public class Solution {
    public int UniqueMorseRepresentations(string[] words) {
        string[] sa = {".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};
        var ss = new HashSet<string>();
        
        foreach (var s in words) {
            string ns = "";
            
            foreach (char c in s) {
                ns += sa[c - 'a'];
            }
            ss.Add(ns);
        }
        return ss.Count;
    }
}

//Second fastest
public class Solution {
    public int UniqueMorseRepresentations(string[] words) {
                    var morsesCode = new List<string>() { ".-", "-...", "-.-.", "-..", ".", "..-.",
                "--.", "....", "..", ".---", "-.-",
                ".-..", "--", "-.", "---", ".--.", "--.-", ".-.",
                "...", "-", "..-", "...-", ".--",
                "-..-", "-.--", "--.." };
            var dictionary = new Dictionary<char, string>();
            char letter = 'a';
            foreach (var item in morsesCode)
            {
                dictionary.Add(letter, item);
                letter++;
            }

            var wordsInMoses = new List<string>();
            foreach (var item in words)
            {
                string wordInMoses = "";
                foreach (var c in item)
                {
                    wordInMoses += dictionary[c];
                }
                wordsInMoses.Add(wordInMoses);
            }
            return wordsInMoses.Distinct().Count();
    }
}

//Fastest
public class Solution {
    public int UniqueMorseRepresentations(string[] words) {
        var morseCode = new string[]{".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."};
        var set = new HashSet<string>();
        for(int i=0; i<words.Length; i++) {
            var builder = new StringBuilder();
            for(int k=0; k<words[i].Length; k++) {
                builder.Append(morseCode[words[i][k]-'a']);
            }
            set.Add(builder.ToString());
        }
        return set.Count;
    }
}

File: Solutions/804_Unique_Morse_Code_Words/804_Unique_Morse_Code_Words.js
//Fourth fastest
//https://leetcode.com/problems/unique-morse-code-words/discuss/652411/JavaScript-concat-fcn-to-build-strings-store-in-set
let uniqueMorseRepresentations = (words) => {
    const ca = 'a'.charCodeAt(0);
    let morse = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."];
    let record = new Set();
    for (let i in words){
        let trans = "";
        for (let c of words[i])
            trans = trans.concat(morse[c.charCodeAt(0)-ca]);
        record.add(trans);
    }
    return record.size;
};

//Third fastest
/**
 * @param {string[]} words
 * @return {number}
 */
const createDict = () => {
    const morseAlphabet = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."];
    const alphabet = new Array(26).fill(1).map((_, ind) => String.fromCharCode(ind+97));
    return alphabet.reduce((prev, curr, i) => ({...prev, [alphabet[i]]: morseAlphabet[i]}), {});
}

const dict = createDict();

const parseToMorse = (raw) => [...raw].reduce((prev, curr) => prev + dict[curr], '');

const uniqueMorseRepresentations = (words) => {
    return new Set(words.reduce((prev, curr) => [...prev, parseToMorse(curr)], [])).size;
};

//Second fastest
/**
 * @param {string[]} words
 * @return {number}
 */
var uniqueMorseRepresentations = function(words) {
    const ca = 'a'.charCodeAt(0);
   let morse = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."];
   let record = new Set();
   for (let i in words){
       let trans = "";
       for (let c of words[i])
           trans = trans.concat(morse[c.charCodeAt(0)-ca]);
       record.add(trans);
   }
   return record.size;
};

//Fastest
const morseCodes = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."];
const letters = 'abcdefghijklmnopqrstuvwxyz'.split('');
const morse = {};
letters.forEach((letter, i) => {
   morse[letter] = morseCodes[i];
});
/**
 * @param {string[]} words
 * @return {number}
 */
var uniqueMorseRepresentations = function(words) {
    const result = {};
    for (const word of words) {
        const mapped = [];
        for (let i = 0; i < word.length; i++) {
            mapped.push(morse[word[i]]);
        }
        result[mapped.join('')] = 1;
    }
    return Object.keys(result).length;
};